composer/trainer/trainer.py

```diff
@@ -1652,15 +1663,21 @@ def _train_batch(self, use_grad_scaling: bool):
                             total_loss = self.state.scaler.step(
                                 optimizer, closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs))
                         else:
-                            total_loss = optimizer.step(
-                                closure=lambda **kwargs: self._train_microbatches(microbatches, **kwargs).item())
+                            if type(self._device) == DeviceTPU:
+                                total_loss = xm.optimizer_step(optimizer)
```

Comment:
I'm not thrilled about device-specific code in the training loop. Thoughts about adding a `step(self, optimizer)` onto the `Device` class? Also not thrilled about this alternative, as then the code would be the same for cpu and gpu. IDK, thoughts?

Reply:
I think since some code is device specific in a few places + otherwise we'd have to right the same code for cpu, gpu, this is fine for now. As we add support for different types of hardware we can refactor.